src/logic/map.cc

```diff
@@ -540,154 +540,291 @@ void Map::set_origin(const Coords& new_origin) {
+				was_preserved[get_index(co)] = true;
+			} else {
+				Field& field = new_fields[get_index(cn, w)];
 				field.set_height(10);
 				field.set_terrains(default_terrains);
 			}
 		}
 	}
+	for (Coords& c : starting_pos_) {
+		if (c) {  // only if set (!= Coords::null())
+			c = lambda_cn(c);
+		}
+	}
 
-	// Replace all fields
-	fields_.reset(new Field[field_size]);
-	clear_array<>(&fields_, field_size);
-	for (size_t ind = 0; ind < field_size; ++ind) {
-		fields_[ind] = new_fields[ind];
+	{
+		PortSpacesSet new_port_spaces;
+		for (const Coords& c : port_spaces_) {
+			if (Coords cn = lambda_cn(c)) {
+				new_port_spaces.insert(cn);
+			}
+		}
+		port_spaces_ = new_port_spaces;
+	}
+
+	for (int16_t x = 0; x < width_; ++x) {
+		for (int16_t y = 0; y < height_; ++y) {
+			Coords c(x, y);
+			if (!was_preserved[get_index(c)]) {
+				Field& f = operator[](c);
+				if (upcast(Immovable, i, f.get_immovable())) {
+					i->remove(egbase);
+				}
+				while (Bob* b = f.get_first_bob()) {
+					b->remove(egbase);
+				}
+			}
+		}
 	}
-	log("Resized map from (%d, %d) to (%u, %u) at (%d, %d)\n", width_, height_, w, h, split.x,
-	    split.y);
+
 	width_ = w;
 	height_ = h;
+	fields_.reset(new_fields.release());
+	// Always call allocate_player_maps() while changing the map's size. Forgetting to do so will
+	// result in random crashes.
+	egbase.allocate_player_maps();
+	recalc_whole_map(egbase);
 
-	// Inform immovables and bobs about their new position
-	for (MapIndex idx = 0; idx < field_size; ++idx) {
-		Field& f = operator[](idx);
-		if (upcast(Immovable, immovable, f.get_immovable())) {
-			immovable->position_ = get_fcoords(f);
-		}
-		// Ensuring that all bob iterators are changed correctly is a bit hacky, but the more obvious
-		// solution of doing it like in set_origin() is highly problematic here, or so ASan tells me
-		std::vector<Bob*> bobs;
-		for (Bob* bob = f.get_first_bob(); bob; bob = bob->get_next_bob()) {
-			bobs.push_back(bob);
-		}
-		f.bobs = nullptr;
-		for (Bob* bob : bobs) {
-			bob->position_.field = nullptr;
-			bob->linknext_ = nullptr;
-			bob->linkpprev_ = nullptr;
-			bob->set_position(egbase, get_fcoords(f));
+	for (int16_t x = 0; x < width_; ++x) {
+		for (int16_t y = 0; y < height_; ++y) {
+			Field& f = operator[](Coords(x, y));
+			if (upcast(Immovable, imm, f.get_immovable())) {
+				imm->position_ = get_fcoords(f);
+			}
+			if (Bob* b = f.get_first_bob()) {
+				b->linkpprev_ = &f.bobs;
+			}
+			for (Bob* b = f.get_first_bob(); b; b = b->get_next_bob()) {
+				b->position_.x = x;
+				b->position_.y = y;
+				b->position_.field = &f;
+			}
 		}
 	}
 
+	log("Map was resized to %d×%d\n", width_, height_);
+}
+
+ResizeHistory Map::dump_state(const EditorGameBase&) const {
+	ResizeHistory rh;
+	rh.size.w = width_;
+	rh.size.h = height_;
+	rh.port_spaces = port_spaces_;
+	rh.starting_positions = starting_pos_;
+	for (MapIndex i = max_index(); i; --i) {
+		rh.fields.push_back(FieldData(operator[](i - 1)));
```

Comment:
Why using an opeator[] in such a way cant you use it plain? 
is this not the same as *this[]? Could  we use an operator[x, y] ?

Reply:
I'm just using the same syntax here as everywhere else in the code. using `this` would require extra parenthesis (`FieldData((*this)[i-1])`) which is poorly readable IMHO. Using x/y coords instead of MapIndex would unnecessarily introduce nested `for` loops